cmw.go

```diff
@@ -1,223 +1,263 @@
-func (o *CMW) SetValue(v []byte)         { _ = o.val.Set(v) }
-func (o *CMW) SetIndicators(indicators ...Indicator) {
-	var v Indicator
-
-	for _, ind := range indicators {
-		v.Set(ind)
+func (o CMW) GetKind() Kind { return o.kind }
+func (o CMW) GetFormat() Format {
+	switch o.kind {
+	case KindMonad:
+		return o.monad.format
+	case KindCollection:
+		return o.collection.format
+	default:
+		return FormatUnknown
 	}
-
-	o.ind = v
 }
-func (o *CMW) SetSerialization(s Serialization) { o.serialization = s }
 
-func (o CMW) GetValue() []byte                { return o.val }
-func (o CMW) GetType() string                 { return o.typ.String() }
-func (o CMW) GetIndicator() Indicator         { return o.ind }
-func (o CMW) GetSerialization() Serialization { return o.serialization }
+type Format uint
 
-// Deserialize a CMW
-func (o *CMW) Deserialize(b []byte) error {
-	s := sniff(b)
-
-	o.serialization = s
+const (
+	FormatUnknown = Format(iota)
+	// JSON formats
+	FormatJSONRecord
+	FormatJSONCollection
+	// CBOR formats
+	FormatCBORRecord
+	FormatCBORCollection
+	FormatCBORTag
+)
 
-	switch s {
-	case JSONArray:
-		return o.UnmarshalJSON(b)
-	case CBORArray, CBORTag:
-		return o.UnmarshalCBOR(b)
+func (o Format) String() string {
+	switch o {
+	case FormatJSONRecord:
+		return "JSON record"
+	case FormatJSONCollection:
+		return "JSON collection"
+	case FormatCBORRecord:
+		return "CBOR record"
+	case FormatCBORCollection:
+		return "CBOR collection"
+	case FormatCBORTag:
+		return "CBOR tag"
+	case FormatUnknown:
+		fallthrough
+	default:
+		return "unknown"
 	}
-
-	return errors.New("unknown CMW format")
 }
 
-// Serialize a CMW according to its provided Serialization
-func (o CMW) Serialize() ([]byte, error) {
-	s := o.serialization
-	switch s {
-	case JSONArray:
-		return o.MarshalJSON()
-	case CBORArray, CBORTag:
-		return o.MarshalCBOR()
-	}
-	return nil, fmt.Errorf("invalid serialization format %d", s)
+func NewMonad(mediaType any, value []byte, indicators ...Indicator) *CMW {
+	var c CMW
+	c.val.Set(value)
+	c.typ.Set(mediaType)
+	c.setIndicators(indicators...)
+	c.kind = KindMonad
+	return &c
 }
 
-func (o CMW) MarshalJSON() ([]byte, error) { return arrayEncode(json.Marshal, &o) }
+func (o CMW) GetMonadType() string         { return o.monad.getType() }
+func (o CMW) GetMonadValue() []byte        { return o.monad.getValue() }
+func (o CMW) GetMonadIndicator() Indicator { return o.monad.getIndicator() }
+func (o *CMW) UseCBORTagFormat()           { o.monad.format = FormatCBORTag }
 
-func (o CMW) MarshalCBOR() ([]byte, error) {
-	s := o.serialization
-	switch s {
-	case CBORArray:
-		return arrayEncode(cbor.Marshal, &o)
-	case CBORTag:
-		return o.encodeCBORTag()
+// NewCollection instantiate a new Collection CMW with the supplied __cmwc_t
+// Pass an empty string to avoid setting __cmwc_t
+func NewCollection(cmwct string) *CMW {
+	if cmwct != "" && isValidCollectionType(cmwct) != nil {
+		return nil
```

Comment:
Might be good to allow returning an error here to not lose the info from `isValidCollectionType`.